inspirehep/modules/workflows/views.py

```diff
@@ -376,3 +401,180 @@ def robotupload_callback():
         responses[recid] = response
 
     return jsonify(responses)
+
+
+def _validate_workflow_payload(keys):
+    """Validate and get the workflow data from the request.
+
+    It will first check if all the given keys are part of
+    the request's payload, after will validate the ``metadata`` against
+    the ``hep`` JSONSchema and finally it will pass the ``workflow_data``
+    kwarg.
+
+    Args:
+        keys (list(str)): the keys to check if they exist.
+
+    Raises:
+        CallbackMalformedError: if any of the ``keys`` is missing from
+            the request payload.
+        CallbackValidationError: if the workflow ``metadata`` is not valid
+            against ``hep`` JSONSchema.
+    """
+    workflow_data = request.get_json()
+
+    # Check for missing keys
+    for key in keys:
+        if workflow_data.get(key) is None:
+            raise CallbackMalformedError(key)
+
+    # Check for validation errors
+    try:
+        validate(schema='hep', data=workflow_data['metadata'])
+    except ValidationError:
+        workflow_data['_extra_data']['validation_errors'] = \
+            get_validation_errors(workflow_data['metadata'], 'hep')
+        workflow_data['_extra_data']['callback_url'] = \
+            resolve_validation_callback_url()
+        raise CallbackValidationError(workflow_data)
+    return workflow_data
+
+
+def error_handler(f):
+    """Decorator to handle callback exceptions."""
+
+    @wraps(f)
+    def inner(*args, **kwargs):
+        try:
+            return f(*args, **kwargs)
+        except CallbackMalformedError as error:
+            return make_response(
+                dict(error_code=error.error_code,
+                     message=error.description),
+                code=error.code
+            )
+        except CallbackWorkflowNotFoundError as error:
+            return make_response(
+                dict(error_code=error.error_code,
+                     message=error.description),
+                code=error.code
+            )
+        except CallbackValidationError as error:
```

Comment:
why not use simpler error handlers, per-error instead of per-view?:

    http://flask.pocoo.org/docs/0.12/patterns/apierrors/

I know that right now it works the same, but if we are going to reuse them, might be easier to extend, test and maintain.

Reply:
Incidentally it also simplifies the `make_respoonse` ;)

Reply:
The problem with adding the error handling to the view, is that all the other endpoints are not proper RESTful, it needs refactoring for having common error and message formats.

Reply:
I'm not sure how REST is related to this, but if we start doing it, we can refactor while we add new views as needed (that to me sounds better than doing a big plan beforehand, less risk involved, and you ensure that you are building just what you need).